x/service/keeper/msg_server_add_service_test.go

```diff
@@ -107,26 +126,67 @@ func TestMsgServer_AddService(t *testing.T) {
 		},
 		{
 			desc:          "invalid - service already exists (same service supplier)",
+			preFn:         func(t *testing.T) {},
 			address:       addr,
 			service:       preExistingService,
 			expectedError: types.ErrServiceAlreadyExists,
 		},
 		{
 			desc:          "invalid - service already exists (different service supplier)",
+			preFn:         func(t *testing.T) {},
 			address:       sample.AccAddress(),
 			service:       preExistingService,
 			expectedError: types.ErrServiceAlreadyExists,
 		},
+		{
+			desc:          "invalid - no spendable coins",
+			preFn:         func(t *testing.T) {},
+			address:       sample.AccAddress(),
+			service:       svc1,
+			expectedError: types.ErrServiceNotEnoughFunds,
+		},
+		{
+			desc: "invalid - insufficient upokt balance",
+			preFn: func(t *testing.T) {
+				// Add 999999999 upokt to the account (one less than AddServiceFee)
+				keepertest.AddAccToAccMapCoins(t, validAddr2, "upokt", oneUPOKTGreaterThanFee-2)
+			},
+			address:       validAddr2,
+			service:       svc1,
+			expectedError: types.ErrServiceNotEnoughFunds,
+		},
+		{
+			desc: "invalid - account has exactly AddServiceFee",
+			preFn: func(t *testing.T) {
+				// Add 10000000000 upokt to the account
+				keepertest.AddAccToAccMapCoins(t, validAddr2, "upokt", oneUPOKTGreaterThanFee-1)
+			},
+			address:       validAddr2,
+			service:       svc1,
+			expectedError: types.ErrServiceNotEnoughFunds,
+		},
+		{
+			desc: "invalid - sufficient balance of different denom",
+			preFn: func(t *testing.T) {
+				// Adds 10000000001 wrong coins to the account
+				keepertest.AddAccToAccMapCoins(t, validAddr2, "wrong", oneUPOKTGreaterThanFee)
+			},
+			address:       validAddr2,
+			service:       svc1,
+			expectedError: types.ErrServiceNotEnoughFunds,
+		},
 	}
 
 	for _, tt := range tests {
 		t.Run(tt.desc, func(t *testing.T) {
+			tt.preFn(t)
```

Comment:
This is a cool pattern, but I'm a bit worried about the implicit expectation of order without parallelizability or other bugs that could happen.

Why don't we need a cleanup function that would be called `t.Cleanup(tt.teardown)`? If we do this, I think renaming `preFn` to `setup` just do be a bit more idiomatic with testing lingo.

Reply:
We dont need a cleanup bc the method called in the `setup` method has it's own cleanup as a helper which is called when the `t.Run` closes.